### PR TITLE
SDK onboarding

### PR DESCRIPTION
**Pull request by mshaban-msft**

### CODE REVIEW

File: SDK-Onboarding/readme.md [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+- [Service Release](#service-release)
+- [Understanding the Service](#understanding-the-service)
+- [Development workflow](#development-workflow)
+- [Design Phase (LLC or Conveniece Layer)](#design-phase)
+- [Client generation](#client-generation) (SUPER IMPORTANT!)
+- [Samples](#samples)
+- [Testing](#testing)
+- [Final touches](#final-touches)
+- [SDK release](#sdk-release)
+- User studies
```
Comment (kristapratico) [inline SDK-Onboarding/readme.md:17]: Do you plan to add info about User Studies or should this be removed for now? You could link [here](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/279/User-Experience-Research) which has a getting started and talks about how to form good hypotheses.

File: SDK-Onboarding/readme.md [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+
+### Language-specific Instructions
+- [Python-specific instructions](#python-specific-instructions)
+- [.NET-specific instructions](#net-specific-instructions)
+
+
+
+## Service Release
+
+At this stage, the service team should release the service swagger definitions to [azure-rest-api-specs](https://github.com/Azure/azure-rest-api-specs) repo, and they should also provide at least a working PPE env for us to use for SDK development.
```
Comment (kristapratico) [inline SDK-Onboarding/readme.md:33]: Might be worth describing what a PPE is and how this is not a stable environment

File: SDK-Onboarding/readme.md [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+
+The [azure-rest-api-specs](https://github.com/Azure/azure-rest-api-specs) repo also has CI checks to make sure swagger files submitted are valid syntactically and complies with OpenApi standards (a requirement for [Autorest]([azure-rest-api-specs](https://github.com/Azure/azure-rest-api-specs)) client generation).
```
Comment (kristapratico) [inline SDK-Onboarding/readme.md:35]: Fixes the link

```suggestion
The [azure-rest-api-specs](https://github.com/Azure/azure-rest-api-specs) repo also has CI checks to make sure swagger files submitted are valid syntactically and complies with OpenApi standards (a requirement for [Autorest](https://github.com/Azure/azure-rest-api-specs) client generation).
```

File: SDK-Onboarding/readme.md [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+There are practices for preview and GA swagger releases on [azure-rest-api-specs](https://github.com/Azure/azure-rest-api-specs). For example, preview swaggers are released to main branch, while GA swaggers are released to main branch.
```
Comment (kristapratico) [inline SDK-Onboarding/readme.md:37]: > For example, preview swaggers are released to main branch, while GA swaggers are released to main branch

This describes the same practice for both. Is this true? Maybe it makes sense to describe some of the branching strategy that @heaths has been working on with the service teams.

File: SDK-Onboarding/readme.md [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+
+One of the issues the CI has at the moment, is that there might be discrepancies between the actual service APIs and the published swagger. The team is currently working on some validations to make sure that never happens.
```
Comment (kristapratico) [inline SDK-Onboarding/readme.md:39]: This seems like more of a service issue than a CI issue. Maybe reword to something like..?

> Sometimes there might be discrepancies between the actual service APIs and the published swagger. The team is currently working on some CI validations to make sure that never happens.

File: SDK-Onboarding/readme.md [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+
+## Understanding the Service
+
+At this stage, we try to familiarize ourselves with the service swagger and APIs. We usually meet with someone from the service team. They explain how the service works, walk us through newly added features, breaking changes, and provide postman collection we can use for experimenting with the service.
```
Comment (kristapratico) [inline SDK-Onboarding/readme.md:43]: Just a few thoughts...

Especially with "new to us" services, I think we oftentimes have an arch board introduction to the service by the service team as well. This way we can understand the service champion scenarios and make sure we optimize client library design for common user cases.

The postman collection thing might be service-team specific unless we are pushing for that as a standard now.

File: SDK-Onboarding/readme.md [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+
+We then start experimenting with the APIs using postman, making sure we fully understand the Ins and Outs od the service.
```
Comment (kristapratico) [inline SDK-Onboarding/readme.md:45]: nit spelling

```suggestion
We then start experimenting with the APIs using postman, making sure we fully understand the Ins and Outs of the service.
```

File: SDK-Onboarding/readme.md [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+
+Usually the service should be stable, having already been tested by the service team through multiple bug bashes.
```
Comment (kristapratico) [inline SDK-Onboarding/readme.md:47]: By stable do you mean GA? Or do you mean that the service is deployed to prod regions?

File: SDK-Onboarding/readme.md [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+
+## Development workflow
+Like many open-source projects, development workflow goes as follows
+- Fork the SDK repo for the language you're developing for (look at [resources](#resources) for more info)
+- Clone your repo locally
```
Comment (kristapratico) [inline SDK-Onboarding/readme.md:52]: You might mention to clone your fork specifically

File: SDK-Onboarding/readme.md [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+- Creat a feature branch (in your repository)
```
Comment (kristapratico) [inline SDK-Onboarding/readme.md:53]:
```suggestion
- Create a feature branch (in your repository)
```

File: SDK-Onboarding/readme.md [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+- Develop SDK and commit changes
+- Create a PR from your branch in your repo to Azure SDK main branch
```
Comment (kristapratico) [inline SDK-Onboarding/readme.md:55]: maybe clarify Azure SDK language repo main branch (to not confuse with our azure-sdk repo)

File: SDK-Onboarding/readme.md [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+
+## Design Phase
+At this stage, we decide on how we should proceed with SDK design. we have 2 options to go with
+- LLC client
```
Comment (kristapratico) [inline SDK-Onboarding/readme.md:59]: I think we are calling LLC -> DPG now (at least in .NET, Python, and Java). Might be RLC for JS/TS. So maybe the options should be "Pure DPG library" or "DPG + customizations"?

File: SDK-Onboarding/readme.md [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+- Client with a conveniece layer
```
Comment (kristapratico) [inline SDK-Onboarding/readme.md:60]: nit spelling (a few instances on this page)

```suggestion
- Client with a convenience layer
```

File: SDK-Onboarding/readme.md [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+
+We factor-in things like
+- How many APIs are there
+- Customization level we need
+- Previous user studies (if any)
+
+### LLC client
+This stands for Low-Level Client. This SDK is purely generated. We use generation parameters and directives to customize the SDK.
```
Comment (kristapratico) [inline SDK-Onboarding/readme.md:68]: We have kind of evolved from LLC. I think we should describe DPG here and using patch to customize the code. We could link or include some information from the DPG docs here:

https://github.com/Azure/autorest.python/wiki/Generating-with-DPG

https://github.com/Azure/autorest.python/blob/autorestv3/docs/customizations.md

File: SDK-Onboarding/readme.md [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+
+## Final Touches
+At this stage, you need to privde final touches
+
+### Changelog.md
+Update changelog.md file to add entry for the currnt release
+with things like version, date, new features, and breaking changes (if any).
+
+### Readme.md
+This will be the entrypoint for all developers looking for your SDK. You should add things like an overview of what the service does and some samples.
```
Comment (kristapratico) [inline SDK-Onboarding/readme.md:131]: Maybe link to our general guidelines about documentation: https://azure.github.io/azure-sdk/general_documentation.html

File: SDK-Onboarding/readme.md [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+
+## SDK Release
+At this stage we trigger release pipeline on Azure DevOps to release a package for users (pip package for python, or nuget package for .net).
+
+Check with senior people from SDK team (because you most likely won't have permissions to carry out this step).
```
Comment (kristapratico) [inline SDK-Onboarding/readme.md:136]: You could link to our Permissions page here: https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/67/Permissions

File: SDK-Onboarding/readme.md [hunk 17]
```diff
@@ -0,0 +1,243 @@
...
+
+## CI/CD Pipeline
+
+### Testing pipeline
+We simply have recorded mode because when submitting a PR or pushing new changes, CI/CD pipeline will only run tests in recorded mode to reduce network traffic. But it also has nightly builds which run tests in live mode. 
+
+### Secrets management
+Every project has a tests.yml file which contains info you need for testing like env variables names used by pipeline.
```
Comment (kristapratico) [inline SDK-Onboarding/readme.md:144]: We use tests.yml if we have static env vars, but it's preferred (if possible) to dynamically create the env vars in the test-resources.json and use those in our tests. 

https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/51/Testing-Guidelines

File: SDK-Onboarding/readme.md [hunk 18]
```diff
@@ -0,0 +1,243 @@
...
+- Azure-Core .Net: https://docs.microsoft.com/en-us/dotnet/api/overview/azure/core-readme
+
+- Azure-Core Python: https://docs.microsoft.com/en-us/python/api/overview/azure/core-readme?view=azure-python
+
+### Coding Guidelines
+- coding guidelines: https://azure.github.io/azure-sdk/general_introduction.html
+
+### Internal Tutorials
+- Internal Tutorials:           https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/346/Azure-SDK-Development-Training
+- Azure SDK Internal Training:  https://github.com/Azure/azure-sdk-pr/tree/master/training
```
Comment (kristapratico) [inline SDK-Onboarding/readme.md:163]:
```suggestion
- Azure SDK Internal Training:  https://github.com/Azure/azure-sdk-pr/tree/main/training   
```

File: SDK-Onboarding/readme.md [hunk 19]
```diff
@@ -0,0 +1,243 @@
...
+python -m venv "path/to/venv"
+```
+
+Install development dependencies
+```bash
+cd "path/to/sdk"
+pip install -r dev_requirements.txt
+```
+
+Install client as package (to run samples and tests)
```
Comment (kristapratico) [inline SDK-Onboarding/readme.md:188]: you might add that using `-e` makes this an editable install so your install will reflect any changes made in the code

File: SDK-Onboarding/readme.md [hunk 20]
```diff
@@ -0,0 +1,243 @@
...
+```bash
+cd "path/to/sdk"
+pip install -e .
+```
+
+#### Running Tests
+We can run tests in 2 modes: live and recorded modes. By default, tests run in recorded mode. You need to add the following file to run tests in live mode
```
Comment (kristapratico) [inline SDK-Onboarding/readme.md:195]: I believe we still support this file as a way to switch on live/recorded mode, but I know in Python we want to move to using the `AZURE_TEST_RUN_LIVE` env var instead. 

You might also include instructions or link to setting up tests to use the test proxy since many of our python libraries have migrated to that now. https://github.com/Azure/azure-sdk-for-python/blob/main/doc/dev/tests.md#write-or-run-tests

**Review by kristapratico — COMMENTED** [review]
Thanks for putting this together! I could mostly only comment on the general/python stuff. I'll let someone from .NET review their specific guidance.

File: SDK-Onboarding/readme.md (hunk 4 above)
Comment (heaths) [reply SDK-Onboarding/readme.md:37]: Previews and GAs should be from the main branch but we can, if needed, ship beta SDKs from a preview swagger branch. What I'm working on is actually a different, git-friendly directory structure - mainly editing files in-place vs. making copies, which doesn't work with `git merge`, `git cherry-pick`, etc.
